File: cpp/0290_wordPattern.cpp
// 2个hash，相同规律代表pattern到string之间存在双射
class Solution {
public:
    bool wordPattern(string pattern, string s) {
        if(pattern.size()>s.size())
            return false;
        map<char,string> m1;
        map<string,char> m2;
        int count=0;
        for(int i=0;i<s.size();i++)
        {
            if(count>=pattern.size())
                return false;
            string str = "";
            while(i<s.size() && s[i]!=' ')
                str+=s[i++];
            if(m1.count(pattern[count]))
            {
                if(m1[pattern[count]]!=str)
                    return false;
            }
            else
                m1.insert({pattern[count],str});
            if(m2.count(str))
            {
                if(m2[str]!=pattern[count])
                    return false;
            }
            else
                m2.insert({str,pattern[count]});
            count++;
        }  
        return count>=pattern.size();//注意：必须把pattern遍历完
    }
};

File: java/0290_wordPattern.java
class Solution {
    public boolean wordPattern(String pattern, String s) {
        if(s.length()<pattern.length())
            return false;
        Map<Character, String> m1 = new HashMap<>();
        Map<String,Character> m2 = new HashMap<>();
        int count = 0;
        for(var i=0;i<s.length();i++)
        {
            if(count>=pattern.length())
                return false;
            String str = "";
            while(i<s.length() && s.charAt(i)!=' ')
                str += s.charAt(i++);
            if(m1.containsKey(pattern.charAt(count)))
            {
                if(!(m1.get(pattern.charAt(count)).equals(str)))
                    return false;
            }
            else
                m1.put(pattern.charAt(count),str);
            if(m2.containsKey(str))
            {
                if(m2.get(str)!=pattern.charAt(count))
                    return false;
            }
            else
                m2.put(str,pattern.charAt(count));
            count++;
        }
        return count>=pattern.length();
    }
}
